<a href="https://colab.research.google.com/github/jeremysb1/data_analysis_projects/blob/main/modeling.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Data Modeling

This project seeks to answer the question:

How many customers does this business have?

I will combine multiple data sources into a single dataset representing a customer data model.

## Part 1: Exploring, extracting, and combining customer data

In [ ]:
import pandas as pd
import numpy as np

sales = pd.read_csv('/content/drive/MyDrive/Data Analysis Projects/Data Modeling/purchases.csv')
print(sales.shape)

(71519, 11)


In [ ]:
sales.isnull().sum()

event_time              0
product_id              0
category_id             0
category_code       16739
brand                5707
price                   0
session_id              0
customer_id         18448
guest_first_name    53071
guest_surname       53071
guest_postcode      53071
dtype: int64

Creating a new column to track guest checkouts, which happen when a customer ID is not provided.

In [ ]:
sales['is_guest'] = sales['customer_id'].isnull()

In [ ]:
sales[sales["is_guest"] & sales["customer_id"].notnull()]
sales[(sales["is_guest"] == False) & sales["customer_id"].isnull()]

Empty DataFrame
Columns: [event_time, product_id, category_id, category_code, brand, price, session_id, customer_id, guest_first_name, guest_surname, guest_postcode, is_guest]
Index: []

The data shows all purchases were made by a guest checkout or by a registered customer.

In [ ]:
sales["is_guest"].value_counts(normalize=True)

is_guest
False    0.742055
True     0.257945
Name: proportion, dtype: float64

The proportion of guest vs. registered user purchases.

In [ ]:
guest_columns = ["guest_first_name", "guest_surname", "guest_postcode"]

unique_guests = sales[guest_columns].drop_duplicates()
print(len(unique_guests))

unique_customers = sales["customer_id"].unique()
cust_total = len(unique_customers) + len(unique_guests)

print(len(unique_guests) / (cust_total - 1))

8301
0.2495640671036017


There are around 25,000 unique customer IDs, which represent registered customers, and another roughly 8,000 unique combinations of guest information, so from purchases alone, I estimate the upper bound of the number of customers to be around 33,000.

In [ ]:
guest_columns = ["guest_first_name", "guest_surname", "guest_postcode", "is_guest"]
guests = sales.loc[sales["is_guest"], guest_columns]
guests = guests.drop_duplicates()
guests.head()

guest_first_name guest_surname guest_postcode  is_guest
0           MICHAEL         MASON        RG497ZQ      True
2              COLE     WILKINSON         SW75TQ      True
3          MOHAMMED      RICHARDS        RG150RE      True
7              KIAN         MILLS        SW332TF      True
13             RUBY          OWEN        PO377YS      True

In [ ]:
non_guests = (pd.DataFrame(sales.loc[sales["customer_id"].notnull(), "customer_id"].unique().astype(int), columns=["customer_id"]))

non_guests.head()

customer_id
0         7466
1        31266
2    534142828
3         1035
4         6985

Combining guest and non-guest data:

In [ ]:
sales_customers = pd.concat([non_guests, guests], axis=0, ignore_index=True)

In [ ]:
new_col_names = ["customer_id", "first_name", "surname", "postcode", "is_guest"]
sales_customers = sales_customers.set_axis(new_col_names, axis=1)

In [ ]:
sales_customers["is_guest"] = sales_customers["is_guest"].fillna(False)

In [ ]:
sales_customers["in_purchase_data"] = True

In [ ]:
for col in ["first_name", "surname"]:
    sales_customers[col] = sales_customers[col].str.lower().str.strip()

sales_customers["postcode"] = sales_customers["postcode"].str.strip()

## Explore CRM data

In [ ]:
crm = pd.read_csv('/content/drive/MyDrive/Data Analysis Projects/Data Modeling/crm_export.csv')
print(crm.shape)
crm.head()

(7825, 5)


customer_id first_name  surname postcode  age
0        29223      Holly   Rogers  LS475RT   12
1        27826     Daniel     Owen   M902XX    5
2         7432    Eleanor  Russell  HR904ZA   34
3         2569      Paige  Roberts  DE732EP   61
4         9195    Matilda    Young  LS670FU   78

In [ ]:
crm.isnull().sum()

customer_id    0
first_name     0
surname        0
postcode       0
age            0
dtype: int64

In [ ]:
crm.groupby("customer_id").size().loc[lambda x: x > 1]

Series([], dtype: int64)

In [ ]:
print(len(crm))
print(len(crm.drop(columns = "customer_id").drop_duplicates()))

7825
7419


The next step is to transform the CRM data to the same schema as the customers from the purchase table, and we also need to enhance the registered customers in the purchase history with details from the CRM data.

In [ ]:
sales_and_crm_customers = sales_customers.merge(crm, on="customer_id", how="left", suffixes=("_sales", "_crm"))
print(len(sales_and_crm_customers))
sales_and_crm_customers.isnull().sum()

33261


customer_id          8300
first_name_sales    24961
surname_sales       24961
postcode_sales      24961
is_guest                0
in_purchase_data        0
first_name_crm      26147
surname_crm         26147
postcode_crm        26147
age                 26147
dtype: int64

In [ ]:
merged_customers_filter = (
    (sales_and_crm_customers["customer_id"].notnull())
      &  ((sales_and_crm_customers["first_name_crm"].notnull())
      |  (sales_and_crm_customers["surname_crm"].notnull()))
)

In [ ]:
sales_and_crm_customers.loc[merged_customers_filter, "in_crm_data"] = True
sales_and_crm_customers.loc[~merged_customers_filter, "in_crm_data"] = False
sales_and_crm_customers["in_crm_data"].value_counts()

in_crm_data
False    26147
True      7114
Name: count, dtype: int64

In [ ]:
sales_and_crm_customers.loc[merged_customers_filter, ["first_name_sales", "surname_sales", "postcode_sales"]] = (
    sales_and_crm_customers.loc[merged_customers_filter, ["first_name_crm", "surname_crm", "postcode_crm"]].values
)

In [ ]:
sales_and_crm_customers = (
    sales_and_crm_customers
    .drop(columns=["first_name_crm", "surname_crm", "postcode_crm"])
    .rename(columns={
        "first_name_sales": "first_name",
        "surname_sales": "surname",
        "postcode_sales": "postcode"
    })
)

sales_and_crm_customers.head()

customer_id first_name surname postcode  is_guest  in_purchase_data   age  \
0       7466.0        NaN     NaN      NaN     False              True   NaN   
1      31266.0     Harley  Palmer  HR250EJ     False              True  33.0   
2  534142828.0        NaN     NaN      NaN     False              True   NaN   
3       1035.0        NaN     NaN      NaN     False              True   NaN   
4       6985.0        NaN     NaN      NaN     False              True   NaN   

  in_crm_data  
0       False  
1        True  
2       False  
3       False  
4       False

The customer data from purchases has been enhanced with CRM data where possible.

In [ ]:
crm_ids_to_add = list(set(crm["customer_id"].unique())
    - set(sales_and_crm_customers["customer_id"].unique()))

print(len(crm_ids_to_add))

711


The difference is customer IDs who appear in the CRM data but not in purchases. The output tells us there are 711 such customers whose details need to be added to our growing customer dataset.

In [ ]:
sales_and_crm_customers = (
    pd.concat([sales_and_crm_customers, crm[crm["customer_id"].isin(crm_ids_to_add)]], axis=0, ignore_index=True)
)

In [ ]:
sales_and_crm_customers["is_guest"] = sales_and_crm_customers["is_guest"].fillna(False)
sales_and_crm_customers["in_purchase_data"] = sales_and_crm_customers["in_purchase_data"].fillna(False)
sales_and_crm_customers["in_crm_data"] = sales_and_crm_customers["in_crm_data"].fillna(True)

sales_and_crm_customers.isnull().sum()

customer_id          8300
first_name          17847
surname             17847
postcode            17847
is_guest                0
in_purchase_data        0
age                 26147
in_crm_data             0
dtype: int64

I now have a customer dataset of all customers who made a purchase, including guest checkouts, as well as data from our CRM system, including customers who exist only in the CRM data and have no recorded purchases.

I now need to repeat this process with data in the customer database, which is structurally similar to the CRM data.

## Exploring The Customer Database

In [ ]:
customers = pd.read_csv('/content/drive/MyDrive/Data Analysis Projects/Data Modeling/customer_database.csv')
print(customers.shape)
customers.head()

(23476, 5)


customer_id first_name   surname postcode  age
0         1641       Rhys  Richards  DE456EZ   45
1        24796     Maisie     Young  SW433XX   16
2        14358     Nathan      King   NW49TU   58
3        15306       Jack     Moore  NW908RR   26
4        24971  Alexander   Roberts  SW500HW   85

In [ ]:
customers.isnull().sum()

customer_id    0
first_name     0
surname        0
postcode       0
age            0
dtype: int64

In [ ]:
for col in ["first_name", "surname"]:
    customers[col] = customers[col].str.lower().str.strip()

customers["postcode"] = customers["postcode"].str.strip()

In [ ]:
customers.groupby("customer_id").size().loc[lambda x: x > 1]


Series([], dtype: int64)

In [ ]:
print("{} rows".format(len(customers)))
unique_customers = customers.drop(columns="customer_id").drop_duplicates()
print("{} unique combination of customers".format(len(unique_customers)))

23476 rows
19889 unique combination of customers


The next step is to merge the customer information into the growing customer data by joining them together. Again, we give the duplicate column names meaningful suffixes to show which table they came from.

In [ ]:
all_customers = sales_and_crm_customers.merge(customers, on="customer_id", how="left", suffixes=("_sales", "_customers"))
all_customers.head()

customer_id first_name_sales surname_sales postcode_sales  is_guest  \
0       7466.0              NaN           NaN            NaN     False   
1      31266.0           Harley        Palmer        HR250EJ     False   
2  534142828.0              NaN           NaN            NaN     False   
3       1035.0              NaN           NaN            NaN     False   
4       6985.0              NaN           NaN            NaN     False   

   in_purchase_data  age_sales  in_crm_data first_name_customers  \
0              True        NaN        False                  eve   
1              True       33.0         True                  NaN   
2              True        NaN        False                  NaN   
3              True        NaN        False                 luca   
4              True        NaN        False                  mia   

  surname_customers postcode_customers  age_customers  
0          richards             HR90PT           45.0  
1               NaN                NaN            NaN  
2               NaN                NaN            NaN  
3            gibson            DE256NH           30.0  
4            rogers            HR662RP           43.0

In [ ]:
merged_customers_filter = (
    (all_customers["customer_id"].notnull())
    &  ((all_customers["first_name_customers"].notnull())
        |  (all_customers["surname_customers"].notnull()))
)

all_customers.loc[merged_customers_filter, "in_customer_data"] = True
all_customers.loc[~merged_customers_filter, "in_customer_data"] = False
all_customers["in_customer_data"].value_counts()

in_customer_data
True     22053
False    11919
Name: count, dtype: int64

In [ ]:
update_filter = (
    (all_customers["in_customer_data"])
    &  (all_customers["first_name_sales"].isnull())
    &  (all_customers["surname_sales"].isnull())
)

all_customers.loc[update_filter, ["first_name_sales", "surname_sales", "postcode_sales", "age_sales"]] = (
    all_customers.loc[update_filter, ["first_name_customers", "surname_customers", "postcode_customers", "age_customers"]].values
)

all_customers = (
    all_customers
    .drop(columns=["first_name_customers", "surname_customers", "age_customers", "postcode_customers"])
    .rename(columns={
        "first_name_sales": "first_name",
        "surname_sales": "surname",
        "age_sales": "age",
        "postcode_sales": "postcode"
    })
)

In [ ]:
customer_ids_to_add = list(set(customers["customer_id"].unique())
    - set(all_customers["customer_id"].unique()))

print(len(customer_ids_to_add))

1423


In [ ]:
all_customers = (
    pd.concat([all_customers, customers[customers["customer_id"].isin(customer_ids_to_add)]], axis=0, ignore_index=True)
)

In [ ]:
all_customers["is_guest"] = all_customers["is_guest"].fillna(False)
all_customers["in_purchase_data"] = all_customers["in_purchase_data"].fillna(False)
all_customers["in_crm_data"] = all_customers["in_crm_data"].fillna(False)
all_customers["in_customer_data"] = all_customers["in_customer_data"].fillna(True)

In [ ]:
all_customers.head()

customer_id first_name   surname postcode  is_guest  in_purchase_data  \
0       7466.0        eve  richards   HR90PT     False              True   
1      31266.0     Harley    Palmer  HR250EJ     False              True   
2  534142828.0        NaN       NaN      NaN     False              True   
3       1035.0       luca    gibson  DE256NH     False              True   
4       6985.0        mia    rogers  HR662RP     False              True   

    age  in_crm_data  in_customer_data  
0  45.0        False              True  
1  33.0         True             False  
2   NaN        False             False  
3  30.0        False              True  
4  43.0        False              True

Above is the final schema of the combined customer data model.

In [ ]:
identified_customers = (
    all_customers[(all_customers["customer_id"].notnull())
                  &  (all_customers["in_purchase_data"])
                  &  ((all_customers["in_crm_data"])
                      |  (all_customers["in_customer_data"]))]
)

guests = all_customers[all_customers["is_guest"]]

customer_ids_not_found = (
    all_customers[(all_customers["customer_id"].notnull())
                  &  (all_customers["first_name"].isnull())
                  &  (all_customers["surname"].isnull())]
)

customer_data_only = (
    all_customers[((all_customers["in_crm_data"])
                   |  (all_customers["in_customer_data"])
                   )
                   &  (all_customers["in_purchase_data"] == False)]
)

print(len(all_customers), len(identified_customers))
print(len(guests), len(customer_ids_not_found), len(customer_data_only))

35395 23713
8300 1248 2134


The output is the following:

- 35,395 records in the entire data model,
- 23,713 of which are identified customers,
- 8,300 guests, 1,248 unidentified customer IDs,
- and 2,134 customers with no purchases.

In [ ]:
print(len(all_customers))
all_customers = all_customers.drop_duplicates()
print(len(all_customers))

35395
35395


In [ ]:
all_customers["customer_id"].agg(["min",  "max"])

min            1.0
max    566239774.0
Name: customer_id, dtype: float64

In [ ]:
all_guests = all_customers[all_customers["is_guest"]].copy()
new_ids = np.arange(-1, -(len(all_guests) + 1), -1)
all_customers.loc[all_customers["is_guest"], "customer_id"] = new_ids

In [ ]:
columns_to_consider = ["first_name", "surname", "postcode"]

duplicates = all_customers[all_customers.duplicated(subset=columns_to_consider, keep=False)]

In [ ]:
duplicate_dict = duplicates.groupby(columns_to_consider)['customer_id'].apply(list).to_dict()

In [ ]:
all_customers['other_customer_ids'] = all_customers.apply(lambda x:  duplicate_dict.get((x['first_name'], x['surname'], x['postcode'])), axis=1)

In [ ]:
def remove_own_record(row):
    ids = list(row["other_customer_ids"])
    ids.remove(row["customer_id"])
    return ids

all_customers.loc[all_customers["other_customer_ids"].notnull(), "duplicate_customer_ids"] = (
    all_customers[all_customers["other_customer_ids"].notnull()].apply(remove_own_record, axis=1)
)

In [ ]:
all_customers["rank"] = all_customers.groupby(columns_to_consider).cumcount()+1

In [ ]:
all_customers.loc[all_customers["rank"] == 1, "is_main"] = True
all_customers["is_main"] = all_customers["is_main"].fillna(False)
all_customers = all_customers.drop(columns="rank")

In [ ]:
print(f"Total customers in DB: {len(all_customers)}")
print(f"Of which {len(all_customers[all_customers['is_main']])} are unique/main records")

Total customers in DB: 35395
Of which 29985 are unique/main records


In [ ]:
!pip install recordlinkage
import recordlinkage

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 926.9/926.9 kB 3.1 MB/s eta 0:00:00


In [ ]:
indexer = recordlinkage.Index()
indexer.block('postcode')
candidate_links = indexer.index(all_customers.set_index("customer_id"))

In [ ]:
compare = recordlinkage.Compare()
compare.string('first_name', 'first_name', method='damerau_levenshtein', threshold=0.85, label='first_name')
compare.string('surname', 'surname', method='damerau_levenshtein', threshold=0.85, label='surname')
compare.exact('postcode', 'postcode', label='postcode')

<Compare>

In [ ]:
compare_vectors = compare.compute(candidate_links, all_customers.set_index("customer_id"))

In [ ]:
matches = compare_vectors[compare_vectors.sum(axis=1) == 3]

In [ ]:
match_df = pd.DataFrame(data=matches.index.tolist(), columns=["customer_id_1", "customer_id_2"])
matched = all_customers.merge(match_df, left_on="customer_id", right_on="customer_id_1", how="left", suffixes=("_customers", "_matches"))
matched = matched.merge(match_df, left_on="customer_id", right_on="customer_id_2", how="left", suffixes=("_customers", "_matches"))

In [ ]:
def merge_duplicates(group):
    duplicate_list = []
    if np.isnan(group["customer_id_1_matches"].values[0]) == False:
        duplicate_list.extend(group["customer_id_1_matches"].tolist())
    if np.isnan(group["customer_id_2_matches"].values[0]) == False:
        duplicate_list.extend(group["customer_id_2_matches"].tolist())
    if len(duplicate_list) > 0:
        return sorted(list(set([int(x) for x in duplicate_list])))
    return np.nan

linkages = (
    matched
    .groupby("customer_id")
    .apply(merge_duplicates)
    .reset_index(name="linked_duplicates")
)


In [ ]:
all_customers = all_customers.merge(linkages, on="customer_id", how="left")

In [ ]:
(
    all_customers[(all_customers["duplicate_customer_ids"].notnull())
               &  (all_customers["linked_duplicates"].notnull())
               &  (all_customers["duplicate_customer_ids"]
  !=  all_customers["linked_duplicates"])]
)

customer_id first_name   surname postcode  is_guest  in_purchase_data  \
1          31266.0     Harley    Palmer  HR250EJ     False              True   
6          28961.0       skye   johnson    M80NA     False              True   
10         12586.0        Max     Moore   M902XX     False              True   
12         22825.0     alicia      wood  SO879UN     False              True   
34          2108.0       leon   thomson  DE624BH     False              True   
...            ...        ...       ...      ...       ...               ...   
33227      -8267.0       jack      hunt   SO44XQ      True              True   
33230      -8270.0        mia   chapman  LS806XB      True              True   
33242      -8282.0      zahra    graham   M560SB      True              True   
33246      -8286.0     maisie      king  SW465QQ      True              True   
33259      -8299.0    michael  harrison  HR280TG      True              True   

        age  in_crm_data  in_customer_data  other_customer_ids  \
1      33.0         True             False   [31266.0, 5411.0]   
6      54.0        False              True  [28961.0, 12140.0]   
10     20.0         True             False  [12586.0, 32914.0]   
12     24.0        False              True  [22825.0, 28495.0]   
34     45.0        False              True   [2108.0, 30218.0]   
...     ...          ...               ...                 ...   
33227   NaN        False             False  [-8267.0, 21683.0]   
33230   NaN        False             False   [-8270.0, 9319.0]   
33242   NaN        False             False  [-8282.0, 11280.0]   
33246   NaN        False             False  [-8286.0, 24936.0]   
33259   NaN        False             False   [-8299.0, 5652.0]   

      duplicate_customer_ids  is_main linked_duplicates  
1                   [5411.0]     True     [5411, 31266]  
6                  [12140.0]     True    [12140, 28961]  
10                 [32914.0]     True    [12586, 32914]  
12                 [28495.0]     True    [22825, 28495]  
34                 [30218.0]     True     [2108, 30218]  
...                      ...      ...               ...  
33227              [21683.0]     True    [-8267, 21683]  
33230               [9319.0]     True     [-8270, 9319]  
33242              [11280.0]     True    [-8282, 11280]  
33246              [24936.0]     True    [-8286, 24936]  
33259               [5652.0]     True     [-8299, 5652]  

[4163 rows x 13 columns]

## Conclusions and recommendations

The initial problem statement asked help counting customers. The final data model has 35,395 rows corresponding to 29,985 unique customers.

There are around 26,000 customer IDs in the purchase data that have no corresponding records in either customer dataset. This means there are 26,000 customers who have signed up to make a purchase online, but I don’t have their details, and because they weren’t using a guest checkout, I have nothing but a customer ID for them in our final data model.

I could choose to drop them as incomplete records, but that would skew the measurement of the size of the customer base. It is better to understand that the data model varies in completeness and is suitable for some tasks – like counting customers – but is not wholly suitable for other tasks, such as customer segmentation.

This is a conclusion I would also share with the stakeholders when presenting an analysis based on the data models.



